File: swift/Sources/AetherMedia/Social/SocialGraph.swift
import Foundation

/// Thread-safe social graph backed by a Swift `actor`.
///
/// Tracks which UHIDs the local user is following.  Because it's an actor
/// all mutations and reads are serialised on the actor's executor, making
/// this safe to use from multiple async contexts simultaneously.
public actor SocialGraph {
    private var following: Set<String> = []

    public init() {}

    /// Add `uhid` to the following set.  No-op if already following.
    public func follow(_ uhid: String) {
        guard !uhid.trimmingCharacters(in: .whitespaces).isEmpty else { return }
        following.insert(uhid)
    }

    /// Remove `uhid` from the following set.  No-op if not following.
    public func unfollow(_ uhid: String) {
        following.remove(uhid)
    }

    /// Returns true when `uhid` is in the following set.
    public func isFollowing(_ uhid: String) -> Bool {
        following.contains(uhid)
    }

    /// Returns a sorted array of all followed UHIDs.
    public func followingList() -> [String] {
        following.sorted()
    }

    /// Number of followed accounts.
    public var count: Int { following.count }
}

File: kotlin/src/main/kotlin/social/SocialGraph.kt
package aether.media.social

import java.util.concurrent.ConcurrentHashMap

/**
 * Thread-safe social graph tracking which UHIDs the local user is following.
 *
 * Uses [ConcurrentHashMap] as a set — all operations are O(1) and lock-free
 * on the read path.
 */
class SocialGraph {
    private val following: MutableSet<String> = ConcurrentHashMap.newKeySet()

    /** Add [uhid] to the following set.  No-op if already following. */
    fun follow(uhid: String) {
        require(uhid.isNotBlank()) { "uhid must not be blank" }
        following.add(uhid)
    }

    /** Remove [uhid] from the following set.  No-op if not following. */
    fun unfollow(uhid: String) {
        following.remove(uhid)
    }

    /** Returns true when [uhid] is in the following set. */
    fun isFollowing(uhid: String): Boolean = following.contains(uhid)

    /** Returns a sorted, immutable snapshot of followed UHIDs. */
    fun getFollowing(): List<String> = following.toSortedSet().toList()

    /** Number of followed accounts. */
    val count: Int get() = following.size
}
